pios: f4: flash_internal: erase flash from RAM

mlyle opened the pull request:
Need to have this one little routine in RAM, so that we can do
lengthy flash operations without hanging long enough for watchdog
to expire.

Still needs to have a way to test arming status at the outset. I'm
thinking probably a weak symbol or something like that (as this code can
be used in the loader, too).

fixes #1482

mlyle committed:
pios: f4: flash_internal: erase flash from RAM

Need to have this one little routine in RAM, so that we can do
lengthy flash operations without hanging long enough for watchdog
to expire.

Still needs to have a way to test arming status at the outset. I'm
thinking probably a weak symbol or something like that (as this code can
be used in the loader, too).

mlyle commented:
(doesn't work according to @yamahabest )

yamahabest commented:
Well, there was the issue that the board was not recognized by Windows after auto-upgrading and rebooting. Even after flashing manually it didn't work.
When going back to the previous working firmware, and skipping the auto-upgrade, and flashing manually (through Rescue) the board was recognized by Windows.

The other issue is that after autotune, and disarming, all ESCs start beeping. I believe the FC on disarm reboots/hangs during reboot?
If I leave the GCS connected during the autotune, the representation of the aircraft gets updated irregularly after disarming, just if the telemetry feed is also affected. After disarming, that is. So that would say that the FC doesn't reboot?

mlyle commented:
The code obviously doesn't work :P

yamahabest commented:
I'm just writing down what I noticed ;-)

mlyle commented:
@yamahabest any chance you can try again?  Seems to work on seppuku, at least...

yamahabest commented:
Sure, can you make Jenkins do a build? Just to rule out any local stuff on my side...
Will take a few days though. Can probably test on Wednesday.

tracernz commented:
jenkins, build artifacts please

mlyle commented:
jenkins, build artifacts please

mlyle committed:
pios: f4: flash from ram: fix return sense

mlyle committed:
aq32: shrink the autotune partition

It was unnecessarily large before

mlyle committed:
pios: f4: erase fromram: move to "ramtext" section

mlyle committed:
manualcontrol: maintain a non-UAVO armed flag

So that pios_flash knows when it's unsafe to do long-lived operations.
Factoring can be improved, etc.

mlyle commented:
Rebased, new commits.  jenkins, build artifacts please

tracernz commented on flight/PiOS/STM32F4xx/pios_flash_internal.c:
```diff
@@ -34,6 +36,9 @@
 #include <stdbool.h>
 
+/* Todo: consider relocating to a good header */
+bool __attribute__((weak)) vehicle_is_armed = false;
```
This is kinda ugly IMO. Would be much better to keep the flash API clean and have the module call into it.

mlyle replied:
To be safe, we have to check after disabling interrupts.

mlyle replied:
Oh, you mean a direct cross-module call from manualcontrol -> flash? But f4 flash internal isn't present on all targets, so we could have a weak call symbol or..

tracernz replied:
Yes? What I'm proposing is making this a static local var, and adding a public method to set it which the module can call.

mlyle replied:
OK, but this file isn't in many targets-- do we stub it out?  what if someone doesn't PIOS_INCLUDE_FLASH_INTERNAL etc

tracernz replied:
ifdef like we do everywhere else?

mlyle replied:
I looked at the pios_sys idea and it's not good because it's also not shared everywhere (would need to duplicate code between arches).  Putting it in pios_flash_internal isn't good because the code in manualcontrol would need two ifdefs-- F4 and PIOS_INCLUDE_FLASH_INTERNAL which IMO is destroying a lot of abstraction.

Can we defer?  I think we need a new module to keep track of state like this, and I feel like we need to get into shipping cadence soon.

tracernz replied:
Yep. Not sure the doxygen thing will pick up Todo (vs. TODO) though. :stuck_out_tongue:

mlyle replied:
Improved the TODO comment :P

mlyle committed:
pios_flash_internal: improve TODO comment

tracernz commented:
Just needs test at this point.

mlyle commented:
Yup--- I've mocked up the AQ32 conditions as best as I can on Revolution, but @yamahabest --- testing on actual AQ32 would really be appreciated.

yamahabest commented:
Yep, will do, this evening.

yamahabest commented:
I have tested this.
The first time during upgrading, the board didn't reboot completely, but after unplugging and replugging, the wizard continued. Tried again, and it worked without problems. First time could have been an anomily on my system though.

ESCs still beep, only once, directly after disarming, FC still rebooting?
During my previous test, the ESCs kept on beeping.

The FC does seem to store the Autotune measurements, as the Autotune Wizard comes up when connecting the FC?

mlyle commented:
I assume the beep is because output PWM stops for the couple seconds while flash is resetting.

yamahabest commented:
Ok, is there a way I could actually check that?

mlyle commented:
Sure, you could connect USB and look at uptime after it saves the tune data.

On Fri, Jun 30, 2017 at 2:56 PM yamahabest <notifications@github.com> wrote:

> Ok, is there a way I could actually check that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/d-ronin/dRonin/pull/1817#issuecomment-312381393>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFjF2dxINmIuCsNs3EfHZlJ0MlXyFtIks5sJW7wgaJpZM4N70sY>
> .
>

yamahabest commented:
Ah, right. Will test that once I get to the FC again, will take a couple of days again...
